src/main.cpp

```diff
@@ -5183,7 +5187,7 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
             LOCK(cs_main);
             State(pfrom->GetId())->fProvidesHeaderAndIDs = !(nLocalServices & NODE_WITNESS) || (nCMPCTBLOCKVersion == 2);
             State(pfrom->GetId())->fPreferHeaderAndIDs = fAnnounceUsingCMPCTBLOCK;
-            State(pfrom->GetId())->fWantCmpctWitness = (nCMPCTBLOCKVersion == 2);
+            State(pfrom->GetId())->fSupportsCmpctWitness = (nCMPCTBLOCKVersion == 2);
```

Comment:
This means that we'd overwrite `fSupportsCmpctWitness` if our peer sends a version 2 message followed by a version 1 message.  If we intend to require a certain ordering of sendcmpct announcements, I think that should be more clearly called out in the BIP (I'll leave a separate comment in bitcoin/bips#423 to indicate the language I suggest improving).  But my initial reaction is that we shouldn't impose this constraint.

Anyway I'm a bit confused on the interaction here between NODE_WITNESS and how we process sendcmpct messages.  Per BIP 152, if we don't support version 2 compact blocks, we must ignore version 2 sendcmpct messages.  But this code doesn't appear to be written this way (ie this processes version 2 sendcmpct messages, even if NODE_WITNESS is off).

In particular, it seems that this code could break if eg it was part of a release (like 0.13.0)  where NODE_WITNESS would not be set, but at some point in the future it had a peer that supported only version 2 compact blocks (as we'd process the version 2 message and think we could communicate with the peer, even though we can't).